<launch>
    <node pkg="ToySLAM" type="loam_mapping_node" name="loam_mapping_node" output="screen">
        <!-- HDL-32E specific parameters -->
        <param name="scan_line" type="int" value="32" />
        <param name="edge_threshold" type="double" value="0.25" />
        <param name="surf_threshold" type="double" value="0.08" />
        <param name="map_resolution" type="double" value="0.3" />
        <param name="scan_period" type="double" value="0.05" />
        <param name="optimization_iterations" type="int" value="10" />
        <param name="use_ring_field" type="bool" value="true" />
        
        <!-- Motion detection parameters -->
        <param name="use_aggressive_motion_detection" type="bool" value="true" />
        <param name="min_motion_threshold" type="double" value="0.05" />
        <param name="forced_motion_interval" type="int" value="20" />
        
        <!-- Keyframe parameters - more sensitive -->
        <param name="keyframe_angle_threshold" type="double" value="0.05" />
        <param name="keyframe_distance_threshold" type="double" value="0.2" />
        <param name="keyframe_time_interval" type="int" value="10" />
        
        <!-- ICP parameters -->
        <param name="icp_fitness_threshold" type="double" value="0.3" />
        
        <!-- General parameters -->
        <param name="min_scan_range" type="double" value="1.0" />
        <param name="max_scan_range" type="double" value="80.0" />
        <param name="mapping_flag" type="bool" value="true" />
        <param name="publish_debug_clouds" type="bool" value="true" />
        <param name="use_laser_scan_lines" type="bool" value="true" />
        <param name="use_sub_maps" type="bool" value="true" />
        <param name="enable_motion_compensation" type="bool" value="true" />
        <param name="save_trajectory" type="bool" value="true" />
        <param name="trajectory_filename" type="string" value="$(find ToySLAM)/taslo_trajectory.txt" />
        <param name="map_frame" type="string" value="map" />
        <param name="odom_frame" type="string" value="odom" />
        <param name="lidar_frame" type="string" value="velodyne" />
    </node>
    
    <node pkg="tf" type="static_transform_publisher" name="world_to_map_broadcaster" args="0 0 0 0 0 0 world map 100" />
    
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find ToySLAM)/rviz/lidar_slam.rviz" required="true" />
    
    <!-- <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="trajectory_server_loam" ns="loam_mapping_node" >
        <param name="target_frame_name" type="string" value="map" />
        <param name="source_frame_name" type="string" value="velodyne" />
        <param name="trajectory_update_rate" type="double" value="10.0" />
        <param name="trajectory_publish_rate" type="double" value="10.0" />
    </node> -->
</launch>